### PR TITLE
Feature "line edit"

### PR DESCRIPTION
**Pull request by isbm**
Sometimes Salt should touch only one line or ensure required expression is there. Inspired by [CFEngine "edit_line" promise](https://docs.cfengine.com/docs/3.5/reference-promise-types-files-edit_line.html), now this feature is here. :-)

You can use it from CLI:

```
salt 'e162.suse.de' file.line /etc/foo.conf 'foo = spam' match='foo\s+=.*' mode=replace
```

Or from state:

``` yaml
/etc/foo.conf:
  file.line:
    - content: foo = salty
    - mode: replace
    - match: 'foo\s+=.*' 
```

An example result:

```
e166:/srv/salt # salt 'e162.suse.de' state.highstate
e162.suse.de:
----------
          ID: /etc/foo.conf
    Function: file.line
      Result: True
     Comment: Changes were made
     Started: 20:16:48.104384
    Duration: 3.303 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -1,7 +1,7 @@
                  -# My config
                  -key = value
                  -
                  -# Another comment
                  -foo = spam
                  -
                  -# The end
                  +# My config+key = value++# Another comment+foo = salty++# The end

Summary
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
```

### CODE REVIEW

**Comment by thatch45** [comment]
Very nice! Thanks!

File: salt/modules/file.py [hunk 1]
```diff
@@ -1115,6 +1116,296 @@ def _mkstemp_copy(path,
...
+    elif not occ:
+        msg = 'no'
+    else:
+        msg = None
+
+    if msg:
+        raise Exception('Found {0} expected occurrences in "{1}" expression'.format(msg, target))
+
+
+def line(path, content, match=None, mode=None, location=None,
```
Comment (lorengordon) [inline salt/modules/file.py]: Parameter `match` is undocumented. If my read of the code is correct, `match` is the pattern to search for when deciding which line(s) to replace. If not defined, then `match` is set to `content`. `match` is only used when `mode` is `replace`.

Comment (isbm) [inline salt/modules/file.py]: Oops, true. Will fix it ASAP! Thanks, @lorengordon

**Comment by iggy** [comment]
You also forgot version annotations (even though I thoroughly complained about it in IRC when you were talking about this feature).

**Comment by lorengordon** [comment]
There appears to be a fair amount of overlap in capability between this and `file.replace`. The most significant new capability would seem to me to be the ability to insert a line before and/or after a specific line in the file. Would it make sense to modify/extend `file.replace`, rather than duplicate much of the capability in a different function? If the parameter semantics are desirable, then `file.line` could wrap `file.replace` (similar to how `file.search` works). 

Another difference is how `file.line` reads the entire file into memory rather than working with a line at a time. However, reading the entire file into memory is a change I was looking to make to `file.replace` to support multi-line regex...

**Comment by isbm** [comment]
@iggy I didn't get that at the time you complained. :) Will add.

**Comment by isbm** [comment]
@lorengordon Yes, I am aware it is taking the entire file into the memory. But for now it is OK-ish, because no one really expects the configuration file more than 1M, which is actually nothing these days even on embedded systems. Normally, you suppose to manage the file from your source. The `file.line` is for a special cases, where you are allowed to manage only one line of some configs without touching the rest.

As for extending `file.replace`, I was considering that before implementing `file.line`. But I don't think this is a good idea, because `file.line` does not _just_ replace, but also adds, deletes and ensures (both combined) lines. So while it surely has overlap in the possibility, as well as `file.replace` does with Augeas module respectively, it is not the same thing as `file.replace` not suppose to add or delete anything. Extending `file.replace` would bring a big confusion why replace does not act like one. In this case `file.line` is directly what CFEngine is doing: taking care of _one_ _line_ of the text file.

I agree this is kind of early version, hence I am not thinking to backport it to the 2015.2, but I am not sure that it needs to be merged to other functionality.

**Comment by lorengordon** [comment]
Oh, sorry, I wasn't worried about taking in the whole file into memory. It makes it much easier to perform these operations. I was just commenting on that detail as an implementation difference between the two functions.

`file.replace` does add, with the flags `append_if_not_found` and `prepend_if_not_found`. And you can *almost* delete a line just by passing in `repl=''`, though granted that's going to result in a blank line rather than a removed line. I believe `file.replace` handles the 'ensure' case effectively as well. Like I was saying, the most distinct new capability seems to me to be the ability to insert a line before/after another line.

Which is why I think most of what you're arguing seems to be desirable semantics of `file.line`, which I agree with as it's simpler to comprehend. I'm just saying that it could be accomplished as a wrapper around `file.replace`. As implemented, there is more code and different implementations that make it more difficult to understand and debug for errors. There is also a greater likelihood of drift/regressions, as bugs that affect both functions may only be found and fixed in one (because that's where the issue is reported). For example, right now `file.line` will throw an error if the file doesn't exist, where `file.replace` has a parameter, `ignore_if_not_found`, to control that behavior. Using routines that avoid clobbering symlinks and hardlinks is another area of interest for me, which many programs do not handle well.

Perhaps at the least, it would make sense to identify common functionality and extract it into internal sub-functions?

**Comment by isbm** [comment]
@lorengordon Yep, it works as designed: to throw an error if you want to _edit_ a line in a file that does not exist. The only problem is that I do not (yet) raise a proper exception. That will be surely fixed.

Well, there is a reason why CFEngine has "replace_patterns" _and_ "edit_line". But when I look at the code of `file.replace`, I just understand that originally it was something else and people wanted to have both. So it grew into a multiline monster that should be renamed to something like `file.edit` (or `file.emacs` — SCR!). For example, prepend can not just insert in front of the file, but also _append_ to the end. Is it a bug or an undocumented feature? A 'MULTILINE' flag?..

The `file.line` supposed to focus on only one line. If there is at least an empty file, of course.

**Comment by lorengordon** [comment]
Heh, I'm hardly saying `file.replace` is perfect. I'm not sure what you mean about prepend appending; there are two different, distinct flags to control that behavior. Though I like your implementation better, with the `location` parameter.

Sometimes you can't control exactly what people do with your code. They are likely to use it in ways you didn't imagine, for use cases you didn't imagine. Choosing to always throw an error just because a file is missing, in what may be a tree of hundreds of states applied to systems you may not fully control, is likely to lead to someone (likely me, because I like the `file.line` semantics) submitting an issue because perhaps they don't care if the file is missing. Perhaps only if the file is present do they need to edit it. Now they have to wrap your code in jinja to check if the file exists first. It makes the state files rather unwieldy. (I'm speaking from experience here, as this is our exact use case.)

**Comment by isbm** [comment]
@lorengordon OK! So:
1. Add version
2. Add a proper exception handling.
3. Add a parameter that would suppress it, once file is not there.
4. Anything else we forgot?

But you also sound like you yourself is not very happy of `file.replace` that is now is a swiss-knife multi-purpose function that can itself be split to many! Oh, it also can do _just_ search... :-)

**Comment by isbm** [comment]
@lorengordon And BTW, many thanks for feedback! Seems like you contributed to `file.replace` and I am trashing it too much... :)

**Comment by lorengordon** [comment]
Well, yeah, `file.replace` has gotten a little unwieldy itself, which is why I think much of this overlap should really go into common sub-functions, rather than one monolithic function or two entirely separate functions.

No ego here, I did what I could to fix specific issues I had with `file.replace`, but I know it's not pretty, and I know that my coding skills are not yet up to refactoring the whole thing. :smile:

As for other nit picky bits to add/change... maybe a handful of test cases, to make sure `file.line` does what is expected (and that future changes don't break it)?

Also, I wasn't entirely sure just from reviewing the code, but how does `file.line` handle a case where there is more than one instance of the line? Is that an error condition for `file.line`? If so, is it a valid use case that should be supported? Imagine this type of configuration file, where we want to manage the line `key = foo`:

```
[section 1]
key = foo

[section 2]
key = foo
```

Lastly, what do you think of changing `ensure` to `present`? I think `present` is more in-line with the semantics in many of salt's other functions. `delete` could be `absent`, as well, for the same reason; though the `delete` action is more intuitive than `ensure`, so I feel less strongly about that one.

**Comment by isbm** [comment]
@lorengordon The test cases are absolutely necessary!

Well, if there are same line multiple times, than if you did not specified after and/or before, then first wins.

OK, if present and absent, then how about insert? The insert suppose to insert if and only if there is no line. If there is similar one then it does not. E.g.:
```
foo = bar
```
Then line `foo = spam` will _not_ be added.

**Comment by lorengordon** [comment]
Hmm, so, insert and present are very similar, but present works as if there were a `force=true` flag to overwrite an existing value?

**Comment by isbm** [comment]
Well, the `file.line` was in mind to do _exactly_ what you ask for. So "present" (or ensure) works that if there is `foo = bar` and you want `foo = spam`, then it will _change_ to it. Because you want to make it sure that (the "replace" mode is assumed in this case). However, if there is no `foo = bar` at all, then `foo = spam` will be inserted. But this is different to if you want insert `foo = spam` and there is already `foo = bar`, then insert operation will not be applied, as there is something that starts from `foo = ...`. In this case you would need to use "replace" mode.

Sometimes you want something like: "Add, if nothing there, skip if something". Hence you have the difference between the insert and ensure. The problem with the terminology of "present" that it does not implies what to do if there is very similar line.

**Comment by isbm** [comment]
BTW, following your _valid_ rant about overlaps, I am going to kill search function in `file.replace` by making it literally unable. There is `file.search` for it (which is just a parametrization of `file.replace` at the moment).

**Comment by lorengordon** [comment]
Ok, I see what you mean. But I don't think 'insert' implies what to do either if there is a very similar line. Linguistically, it could just as well mean *always* insert the line, regardless of whether a similar line is present.

I feel like 'absent' and 'present' could be the primary modes, and that 'replace' and 'insert' are parameters that modify the behavior.

*insert* would become (doesn't change existing line if exists, adds it if it doesn't exist):
```
salt 'minion' file.line /etc/foo.conf content='foo = spam' mode=present replace=false insert=['start' | 'end']
```

*ensure* would become (changes line if it exists, adds it if it doesn't exist):
```
salt 'minion' file.line /etc/foo.conf content='foo = spam' mode=present replace=true insert=['start' | 'end']
```

*replace* would become (changes the line if it exists, does not insert it if it does not):
```
salt 'minion' file.line /etc/foo.conf content='foo = spam' mode=present replace=true insert=None
```

**Comment by isbm** [comment]
@lorengordon Would it work:
1. ensure -> present
2. insert -> provided
3. delete -> absent
4. replace -> replaced :-)

**Comment by lorengordon** [comment]
Heh, if you don't like my suggestion about absent/present modes with parameters for the rest, I'd prefer your original terminology. Really, the only one of your original terms that doesn't convey well what it does is 'ensure'. I've been racking my brain trying to think of a term that conveys either insert or replace but I'm coming up with nothin. I like 'present' because it's stateful and it implies doing whatever you need to do to get the specified content into the file, which would mean either insert or replace. 

But when in context with the rest of the options (present, insert, delete, and replace), 'present' feels out of place because it's the only one that's stateful. The rest are actions or tasks. Maybe I need to study CFEngine to see how they handle it...

**Comment by isbm** [comment]
@lorengordon  no no, don't get me wrong, I am open for the proposals. Just trying to understand it entirely. Mea culpa, I am not English native, so I thought it would make sense to use "ensure" because it is a verb, same as others. The "present" or "absent" are either adjective or noun (OK, "absent" is also verb, but not in this case). So I tried to unify these so to speak states. It really makes sense you say present/absent, but should I "fix" the others too or it does not matter if it is noun or verb or adjective?

**Comment by lorengordon** [comment]
Gotcha. I understand about the verbs. I only meant to say that my hangup with "ensure" is that it's not very intuitive for conveying a mode that includes both actions replace or insert. However, neither can I think of a verb that does.

The use of stateful modes, 'absent' and 'present', from a design or user experience perspective, leads me to the parameterization of the other actions (parameters that modify the behavior of 'present'). The use of default values would keep it simple for the majority of use cases (replace=true, insert='end'). The function definition and associated docstring would look like this:

```python
def line(path, content, mode, match=None, replace=True, insert='end',
         before=None, after=None, show_changes=True, backup=False, quiet=False):
    '''
# ... </snip>
    :param mode:
        Present
            This mode is used to add or modify a line in a file. Parameters ``replace`` and ``insert`` control the behavior of this mode.
        Absent
            This mode is used to delete a line in a file.
    :param replace:
        True
            If the line exists, it will be replaced with ``content``. This is the default.
        False
            If the line exists, the existing value will not be replaced.
    :param insert:
        End
            If the line does not exist, add it to the end of the file. This is the default.
        Start
            If the line does not exist, add it to the beginning of the file.
        None
            If the line does not exist, do nothing.
# ... </snip>
    '''
```

**Comment by lorengordon** [comment]
Perhaps we should get some input from others to see what they think? I see @jfindlay on this PR and @cachedout on the newer PR that's still open...

**Comment by jfindlay** [comment]
It doesn't really matter what words are used as long as they are unambiguous and the functions are clearly documented, but it is also important to be consistent with the terminology used by other salt modules and states.

**Comment by isbm** [comment]
@jfindlay Sounds like problem solved, therefore. :)

**Comment by lorengordon** [comment]
If the terminology doesn't matter, then I'll withdraw the question/proposal. I just wanted to discuss it a bit while it's still in the develop branch, as it will be harder to change the parameter names once it's in a release.

**Comment by JensRantil** [comment]
Hi, just some general feedback here as I've tried multiple times to use the `file.line` feature and am having issues every time: Seem like this pull request was prematurely merged. The function has multiple "modes" and which parameters should be used with which mode is highly unclear. Also, the documentation does not come with a single example on how to use the different modes, and after finding this pull request I notice that there is not a single unit test that makes sure that neither the module nor the state does what it does. I was at least hoping to find usage examples in some tests... So, there is essentially no other way for me or anyone else to understand this mystery module/state except sit down and read the code. That sucks. Let's do a better job at merging new modules and states next time. See also #27295.

(sorry, for sounding annoyed here, I'm just tired of debugging this for the third time)

**Comment by isbm** [comment]
@JensRantil I reviewed your writings at #27295 and seems like something went wrong in between, since this feature is old enough and was working. I will look into it (tongue behind the cheek: unit tests should be there too). For now, please use `file.append` instead.

Thanks reporting!

**Pull request merged**